zhaquirks/nimly/lock.py

```diff
@@ -25,11 +32,121 @@
 )
 
 
+class NimlyDoorLock(CustomCluster, DoorLock):
+    """Nimly Door Lock cluster."""
+
+    class AttributeDefs(DoorLock.AttributeDefs):
+        """Nimly Door Lock attribute definitions."""
+
+        nimly_last_lock_unlock_source: Final = ZCLAttributeDef(
+            id=0x100,
+            type=t.bitmap32,
+            access="r",
+            is_manufacturer_specific=True,
+        )
+        nimly_last_pin_code: Final = ZCLAttributeDef(
+            id=0x101,
+            type=t.LVBytes,
+            access="r",
+            is_manufacturer_specific=True,
+        )
+
+
+def last_action_source_converter(value: int) -> str:
+    """Extract last action source value."""
+    value_hex = hex(value)[2:].zfill(8)
+    sources = {
+        "00": "Zigbee",
+        "02": "Keypad",
+        "03": "Fingerprint",
+        "04": "RFID",
+        "0a": "Self",
```

Comment:
I think we should consider using `snake_case` for the enum sensor values, so just "zigbee", "rfid", and so on. We should be able to translate the entries by adding them to `strings.json` in HA (manually) then.